flashlight/fl/dataset/SpanDataset.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include "flashlight/fl/dataset/Dataset.h"
+
+#include <vector>
+
+namespace fl {
+
+/**
+ * A view into an underlying dataset with an offset and optional bounded length.
+ *
+ * The size of the `SpanDataset` is either specified for the size of the input dataset
+ * accounting for the offset.
+ *
+ * We have, for example `SpanDataset(ds, 13).get(i) == ds.get(13 + i)`
+ *
+ * Example:
+  \code{.cpp}
+  // Make a datasets
+  auto makeDataset = []() {
+    auto tensor = fl::rand({5, 4, 10});
+    std::vector<Tensor> fields{tensor};
+    return std::make_shared<TensorDataset>(fields);
+  };
+  auto ds = makeDataset();
+
+  // Create two spanned datasets
+  SpanDataset spands1(ds, 2);
+  SpanDataset spands2(ds, 0, 2);
+  std::cout << spands1.size() << "\n"; // 8 
+  std::cout << spands2.size() << "\n"; // 2
+  std::cout << allClose(spands1.get(3)[0], ds->get(5)[0]) << "\n"; // 1
+  std::cout << allClose(spands2.get(1)[1], ds->get(1)[0]) << "\n"; // 1
+  \endcode
+ */
+class SpanDataset : public Dataset {
+ public:
+  /**
+   * Creates a SpanDataset.
+   * @param[in] dataset The underlying dataset.
+   * @param[in] offset The starting index of the new dataset relative to the
+   * underlying dataset.
+   * @param[in] length The size of the new dataset (if -1, uses previous size
+   * minus the offset)
+   */
+  explicit SpanDataset(
+      std::shared_ptr<const Dataset> dataset,
+      const int64_t offset,
+      const int64_t length = -1);
```

Comment:
nit -- maybe rename length `size`? you use that in the docs/I think that's clearer